text/2017-11-02-guards.md

```diff
@@ -0,0 +1,213 @@
+pub fn pin() -> Guard;
+pub unsafe fn unprotected() -> Guard;
+
+// Methods now take a `&Guard` instead of a `&Scope`.
+impl<T> Atomic<T> {
+    pub fn compare_and_set<'g, O>(
+        &self,
+        current: Ptr<T>,
+        new: Ptr<T>,
+        ord: O,
+        _: &'g Guard,
+    ) -> Result<(), Ptr<'g, T>>
+    where
+        O: CompareAndSetOrdering;
+
+    // The same applies to other methods...
+}
+```
+
+## How reference counting works
+
+Each thread participating in epoch-based garbage collection has some heap-allocated
+data associated with it (in struct `Local`). All such `Local`s are connected into a
+linked list and the head pointer of that list is held in the global data (in struct `Global`).
+`Global` also holds the global epoch and the global garbage queue.
+
+Each `Local` holds a `ManuallyDrop<Arc<Global>>`, thus keeping the garbage collector in
+which it resides alive.
+Similarly, each `Handle` and each `Guard` keeps a pointer to the `Local` associated with it.
+Handles and guards are counted using fields `handle_count` and `guard_count` inside `Local`.
+When both counts reach zero, the `Arc<Global>` inside `Local` is immediately dropped
+(using `ManuallyDrop::drop`), and the heap-allocated `Local` is then marked as deleted.
+If that `Local` was holding the last reference to the `Global`, then the
+`Global` is destroyed as well.
+
+## Changes to the `Collector` interface
+
+If we allow `Handle`s to be `Send`, the following code will compile:
+
+```rust
+let c = Collector::new();
+let h = c.handle();
+let guard = h.pin();
+
+thread::spawn(move || {
+    let guard = h.pin();
+});
+```
+
+This is a suspicious, although not necessarily a *wrong* piece of code. However, if we
+allowed this to be possible, `guard_count` would have to be an atomic integer, and
+atomic operations on the counter would unnecessarily slow down pinning.
+For that reason, `Handle` will not be `Send`.
+
+Currently, `Handle::clone` creates a brand new handle that can be passed to another thread
+(see [PR #26](https://github.com/crossbeam-rs/crossbeam-epoch/pull/26)), but with
+non-`Send` handles that cloning behavior is not very useful anymore.
+Moreover, we haven't been able to find a satisfying signature for the `default_handle` method
+(see [PR #28](https://github.com/crossbeam-rs/crossbeam-epoch/pull/28)), but with
+non-`Send` handles there's an easy solution.
+
+`Handle::clone` will be changed so that it returns a new reference to the same handle (just like
+`Thread::clone` returns a new reference to the same thread).
+
+Then we can implement `default_handle` like this:
+
+```rust
+lazy_static! {
+    static ref COLLECTOR: Collector = Collector::new();
+}
+
+thread_local! {
+    static HANDLE: Handle = COLLECTOR.handle();
+}
+
+pub fn default_handle() -> Handle {
+    HANDLE.with(|handle| handle.clone())
+}
+```
+
+This way `default_handle` bears some cost of reference counting, but it should be
+forgivable, as is demonstrated by the benchmarks...
```

Comment:
It seems there's 1ns diff between `epoch::pin()` and `epoch::default_handle().pin()` (12ns vs. 13ns).  How about specifically pointing out this benchmark result here?

Reply:
Done.